sqlanalytics/resource_sql_endpoint.go

```diff
@@ -0,0 +1,171 @@
+type Tags struct {
+	CustomTags []Tag `json:"custom_tags"`
+}
+
+// Tag ...
+type Tag struct {
+	Key   string `json:"key"`
+	Value string `json:"value"`
+}
+
+// EndpointList ...
+type EndpointList struct {
+	Endpoints []SQLEndpoint `json:"endpoints"`
+}
+
+// NewSQLEndpointsAPI ...
+func NewSQLEndpointsAPI(ctx context.Context, m interface{}) SQLEndpointsAPI {
+	return SQLEndpointsAPI{m.(*common.DatabricksClient), ctx}
+}
+
+// SQLEndpointsAPI ...
+type SQLEndpointsAPI struct {
+	client  *common.DatabricksClient
+	context context.Context
+}
+
+// List all SQL endpoints
+func (a SQLEndpointsAPI) List() (l EndpointList, err error) {
+	err = a.client.Get(a.context, "/sql/endpoints", nil, &l)
+	return
+}
+
+// Start ..
+func (a SQLEndpointsAPI) Start(endpointID string, timeout time.Duration) error {
+	err := a.client.Post(a.context, fmt.Sprintf("/sql/endpoints/%s/start", endpointID), nil, nil)
+	if err != nil {
+		return err
+	}
+	return a.waitForRunning(endpointID, timeout)
+}
+
+// Stop ...
+func (a SQLEndpointsAPI) Stop(endpointID string) error {
+	return a.client.Post(a.context, fmt.Sprintf("/sql/endpoints/%s/stop", endpointID), nil, nil)
+}
+
+// Get ...
+func (a SQLEndpointsAPI) Get(endpointID string) (se SQLEndpoint, err error) {
+	err = a.client.Get(a.context, fmt.Sprintf("/sql/endpoints/%s", endpointID), nil, &se)
+	return
+}
+
+// Create ...
+func (a SQLEndpointsAPI) Create(se *SQLEndpoint, timeout time.Duration) error {
+	// maybe response should be something else...
+	err := a.client.Post(a.context, "/sql/endpoints", se, se)
+	if err != nil {
+		return err
+	}
+	return a.waitForRunning(se.ID, timeout)
+}
+
+func (a SQLEndpointsAPI) waitForRunning(id string, timeout time.Duration) error {
+	return resource.RetryContext(a.context, timeout, func() *resource.RetryError {
+		endpoint, err := a.Get(id)
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+		switch endpoint.State {
+		case "RUNNING":
+			return nil
+		case "DELETED":
+			return resource.NonRetryableError(
+				fmt.Errorf("Endpoint got deleted during creation"))
+		default:
+			msg := fmt.Errorf("Endpoint %s is %s", id, endpoint.State)
+			log.Printf("[INFO] %s", msg.Error())
+			return resource.RetryableError(msg)
+		}
+	})
+}
+
+// Edit ...
+func (a SQLEndpointsAPI) Edit(se SQLEndpoint) error {
+	return a.client.Post(a.context, fmt.Sprintf("/sql/endpoints/%s/edit", se.ID), se, nil)
+}
+
+// Delete ...
+func (a SQLEndpointsAPI) Delete(endpointID string) error {
+	return a.client.Delete(a.context, fmt.Sprintf("/sql/endpoints/%s", endpointID),
+		map[string]interface{}{})
+}
+
+// ResourceSQLEndpoint ...
+func ResourceSQLEndpoint() *schema.Resource {
+	s := internal.StructToSchema(SQLEndpoint{}, func(
+		m map[string]*schema.Schema) map[string]*schema.Schema {
+		m["cluster_size"].ValidateDiagFunc = validation.ToDiagFunc(
+			validation.StringInSlice([]string{
+				"2X-Small", "X-Small", "Small", "Medium", "Large", "X-Large", "2X-Large", "3X-Large", "4X-Large",
```

Comment:
I would move this into the constant somewhere at beginning, so it would be easier to change